flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushWatermarkIntoTableSourceScanRuleBase.java

```diff
@@ -117,8 +126,33 @@ protected FlinkLogicalTableSourceScan getNewScan(
                 idleTimeoutMillis = -1L;
             }
 
+            Optional<RelHint> optionsHintOptional =
+                    scan.getHints().stream()
+                            .filter(
+                                    relHint ->
+                                            relHint.hintName.equalsIgnoreCase(
+                                                    FlinkHints.HINT_NAME_OPTIONS))
+                            .findFirst();
+            Configuration hintOptions = new Configuration();
+            if (optionsHintOptional.isPresent()) {
+                RelHint optionsHint = optionsHintOptional.get();
+                hintOptions = Configuration.fromMap(optionsHint.kvOptions);
+            }
+            Configuration tableOptions = new Configuration();
+            RelOptTable table = scan.getTable();
+            if (table instanceof TableSourceTable) {
+                Map<String, String> tableConfigs =
+                        ((TableSourceTable) table)
+                                .contextResolvedTable()
+                                .getResolvedTable()
+                                .getOptions();
+                tableOptions = Configuration.fromMap(tableConfigs);
+            }
```

Comment:
We can also improve the code to avoid creating unnecessary objects  via 
~~~java
  Configuration tableOptions =
          table instanceof TableSourceTable
                  ? Configuration.fromMap(
                          ((TableSourceTable) table)
                                  .contextResolvedTable()
                                  .getResolvedTable()
                                  .getOptions())
                  : new Configuration();
~~~

Reply:
modified